app/services/hts_service/workflow_engine.rb

```diff
@@ -0,0 +1,194 @@
+    }.freeze
+
+    STATE_CONDITIONS = {
+
+      PREGNANCY_STATUS => %i[gender_based_encounter_check?],
+
+      CIRCUMCISION => %i[gender_based_encounter_check?],
+
+      SOCIAL_HISTORY => %i[marital_status_track?],
+
+      SCREENING => %i[workflow_status_track?],
+
+      APPOINTMENT => %i[workflow_status_track?],
+
+      REFERRAL => %i[workflow_status_track?],
+
+      PARTNER_RECEPTION => %i[workflow_status_track?
+                              check_hiv_results_status?],
+
+    }.freeze
+
+    # Concepts
+
+    def load_states (encounter)
+
+      case encounter
+
+        when 'PREGNANCY STATUS'
+            state = "PREGNANCY STATUS"
+        when "CIRCUMCISION"
+           state = "CIRCUMCISION STATUS"
+        when 'SOCIAL HISTORY'
+           state = "SOCIAL HISTORY"
+        when 'SCREENING'
+           state = "SCREENING"
+        when "REFERRAL"
+            state ="REFERRAL"
+        when "HTS Contact"
+            state ="CONTACT"
+        when 'APPOINTMENT'
+            state = "REFERRAL APPOINTMENT"
+        when "Partner Reception"
+             state ='PARTNER RECEPTION'
+        end
+
+       return state
+
+    end
+
+    def next_state(current_state)
+      ENCOUNTER_SM[current_state]
+    end
+
+
+    def encounter_exists?(type)
+
+         @encounter = type
+         Encounter.where(type: type, patient: @patient, program: @program)\
+                  .where('encounter_datetime BETWEEN ? AND ?', *TimeUtils.day_bounds(@date))\
+                  .exists?
+
+    end
+
+    def valid_state?(state)
+
+      return false if encounter_exists?(encounter_type(state))
+
+             (STATE_CONDITIONS[state] || []).all? { |condition| send(condition) }
+    end
+
+    def gender_based_encounter_check?
+
+            status = Encounter.where(type: @encounter, patient: @patient, program: @program).exists?
+            case @encounter.name
+
+                when 'PREGNANCY STATUS'
+                     return true if @patient.gender == "F" && status == false
+
+                when "CIRCUMCISION"
+                     return true if @patient.gender == "M" && status == false
+            end
+    end
+
+    def marital_status_track?
+
+           concept_id = ConceptName.find_by_name('Civil status').concept_id
+           status = Observation.where(concept_id: concept_id, person_id: @patient.id).exists?
+           return true if status == false
+    end
+
+    def workflow_status_track?
+
+      encounter_type = EncounterType.find_by name: @encounter.name
+      encounter = Encounter.where(patient: @patient, type: encounter_type, program: @program)\
+                           .where('encounter_datetime BETWEEN ? AND ?', *TimeUtils.day_bounds(@date))\
+                           .order(encounter_datetime: :desc)\
+                           .first
+      return true if encounter.blank?
+
+    end
+
+     def check_hiv_results_status?
+
+       return true if true == Observation.joins(:encounter)\
```

Comment:
is it necessary to do `return true if true ===...` when `.exists?` returns a boolean?